### ISSUE TITLE
shutdown-sql: sql predicate query under cluster shutdown conditions. 3.8-SNAPSHOT

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#8931]
It look's like its possible to get a duplicate item in the predicate results set, under cluster shutdown conditions. however we do not see the fail on every run, involving multi cluster member shutdowns and restarts. possible only happens is index is used.

query running on 4 types of Map,  HD or heap,  indexed and non-indexed.

Jenkins nightly build
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/

some times Fails
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/9/console

```
HzClient2HZ sql1 hz.map.sql.predicate.PredCheck threadId=3 global.AssertionException: (id>=259 AND id<269) on map mapBak1HDSqlIdx returned 11 expected 10 
HzClient3HZ sql2 hz.map.sql.predicate.PredCheck threadId=0 global.AssertionException: (id>=258 AND id<268) on map mapBak1SqlIdx returned 11 expected 10 
HzClient1HZ sql2 hz.map.sql.predicate.PredCheck threadId=0 global.AssertionException: (id>=812 AND id<822) on map mapBak1SqlIdx returned 11 expected 10 
HzClient1HZ sql1 hz.map.sql.predicate.PredCheck threadId=2 global.AssertionException: (id>=813 AND id<823) on map mapBak1HDSqlIdx returned 11 expected 10 
```

/disk1/jenkins/workspace/shutdown-sql/3.8-SNAPSHOT/2016_09_21-13_37_20

output/HZ/HzClient3HZ/exception.txt
output/HZ/HzClient4HZ/exception.txt
output/HZ/HzClient2HZ/exception.txt
output/HZ/HzClient1HZ/exception.txt

example logging 

cat /disk1/jenkins/workspace/shutdown-sql/3.8-SNAPSHOT/2016_09_21-13_37_20/output/HZ/HzClient1HZ/out.txt

```
mapBak1SqlIdx
Person [firstName=firstName820, id=820, orgId=0, lastName=lastName820, salary=364.8578864385621]
Person [firstName=firstName821, id=821, orgId=3, lastName=lastName821, salary=746.821683095925]
Person [firstName=firstName822, id=822, orgId=1, lastName=lastName822, salary=687.8337246289309]
Person [firstName=firstName823, id=823, orgId=6, lastName=lastName823, salary=805.6364834570859]
Person [firstName=firstName824, id=824, orgId=0, lastName=lastName824, salary=664.4377290854891]
Person [firstName=firstName825, id=825, orgId=8, lastName=lastName825, salary=511.0124856471277]
Person [firstName=firstName826, id=826, orgId=6, lastName=lastName826, salary=475.2057839980638]
Person [firstName=firstName827, id=827, orgId=5, lastName=lastName827, salary=110.78088977805108]
Person [firstName=firstName828, id=828, orgId=0, lastName=lastName828, salary=742.7535050470094]
Person [firstName=firstName829, id=829, orgId=4, lastName=lastName829, salary=13.52459167016932]
Person [firstName=firstName820, id=820, orgId=0, lastName=lastName820, salary=364.8578864385621]
global.AssertionException: (id>=820 AND id<830) on map mapBak1SqlIdx returned 11 expected 10
        at hz.map.sql.predicate.PredCheck.timeStep(PredCheck.java:26)
        at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:61)
        at remote.bench.marker.MetricsMarker.timeBenchFlatOut(MetricsMarker.java:45)
        at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:36)
        at remote.bench.BenchThread.call(BenchThread.java:38)
        at remote.bench.BenchThread.call(BenchThread.java:12)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```

person with ID 820  is repeated in the results set.

Test https://github.com/Danny-Hazelcast/hzCmd-bench/tree/alocation/lab/hz/shutdown/sql

Fill map with persons, using unique id
https://github.com/Danny-Hazelcast/hzCmd-bench/blob/ab18c3e0b0dcaac9e198de66036ad3b607be8d41/src/main/java/hz/map/member/LoadPersons.java

Make the sql query operations, and check the results
https://github.com/Danny-Hazelcast/hzCmd-bench/blob/alocation/src/main/java/hz/map/sql/predicate/PredCheck.java

From the load stage 
https://github.com/Danny-Hazelcast/hzCmd-bench/blob/alocation/lab/hz/shutdown/sql/load

**Comment by Danny-Hazelcast** [comment]
This fail is looking easier to reproduce when the predicate returns a larger result set.

I have increased the number of person objects loaded into the map.  from 1000 to 10000
I have increased the size of the results set returned from 10 to 1000


build https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/16/console

```
HzClient1HZ sql_mapBak1HDSqlIdx hz.map.sql.predicate.PredCheck threadId=2 global.AssertionException: (id>=956 AND id<1956) on map mapBak1HDSqlIdx returned 1004 expected 1000
```
http://54.87.52.100/~jenkins/workspace/shutdown-sql/3.8-SNAPSHOT/2016_09_21-19_13_38/ 
 /disk1/jenkins/workspace/shutdown-sql/3.8-SNAPSHOT/2016_09_21-19_13_38

output/HZ/HzClient1HZ/exception.txt


cat output/HZ/HzClient1HZ/out.txt  | grep ref\(4212770\) | sort | uniq -c | sort
```
      1 ref(4212770) (id>=956 AND id<1956) on map mapBak1HDSqlIdx size=1004 expected=1000
      1 ref(4212770) Person [firstName=firstName1000, id=1000, orgId=9, lastName=lastName1000, salary=692.0928076577607]
...
..
.
  1 ref(4212770) Person [firstName=firstName998, id=998, orgId=2, lastName=lastName998, salary=75.66305631082959]
      1 ref(4212770) Person [firstName=firstName999, id=999, orgId=0, lastName=lastName999, salary=973.888582910965]
      2 ref(4212770) Person [firstName=firstName1057, id=1057, orgId=9, lastName=lastName1057, salary=230.77041214963253]
      2 ref(4212770) Person [firstName=firstName1562, id=1562, orgId=6, lastName=lastName1562, salary=858.2488754677922]
      2 ref(4212770) Person [firstName=firstName1865, id=1865, orgId=5, lastName=lastName1865, salary=811.0653813473158]
      2 ref(4212770) Person [firstName=firstName966, id=966, orgId=4, lastName=lastName966, salary=684.4199437590537]
```

**Comment by Danny-Hazelcast** [comment]
and next run 
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/17/console

```
HzClient3HZ sql_mapBak1SqlIdx hz.map.sql.predicate.PredCheck threadId=2 global.AssertionException: (id>=7500 AND id<8500) on map mapBak1SqlIdx returned 1005 expected 1000 
```

**Comment by Danny-Hazelcast** [comment]
and next build 18 was ok 
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/18/

**Comment by Danny-Hazelcast** [comment]
@enesakar  @jerrinot  who can we assigne this to  give its a prerequisites for roll up 
https://hazelcast.atlassian.net/wiki/display/PM/Rolling+Upgrade+nightly+AWS+Test
and 
https://hazelcast.atlassian.net/wiki/display/PM/%5BSB%5D+Rolling+Member+Upgrades+for+Minor+Releases

**Comment by jerrinot** [comment]
@Danny-Hazelcast can you confirm it's reproducible only when index is enabled?

**Comment by Danny-Hazelcast** [comment]
@jerrinot 

/disk1/jenkins/workspace/shutdown-sql/3.8-SNAPSHOT
find . -name exception.txt | xargs grep -h AssertionException | grep return | sort | uniq

shows that only map with index, have caused the issue so fare.

mapBak1HDSqlIdx 
mapBak1SqlIdx

however sample size is small (only 4 failing runs from 13). 
so its not the same as saying   "reproducible only when index is enabled"

but its a hint

**Comment by Danny-Hazelcast** [comment]
after a few parring run's,  we have a hit this issue again. 
this shows the issue is difficult / rare to reproduce
https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-sql/69/console

**Comment by vbekiaris** [comment]
This issue should be fixed once #8385 is done.

**Comment by Danny-Hazelcast** [comment]
false alarme some how the test was running with very old jars
